---
title: "Running Whippet"
output: html_document
---

Now that we have obtained informative clusters on the interneuron dataset, we can begin to compare the splicing profiles between group of cells. We will try:
1. Compare profiles of proliferative vs non-proliferative interneurons
2. Compare profiles of proliferative states for each cell type (CGE,dMGE,vMGE)
3. Compare profiles of cell types within each proliferative state
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## 1. Retrieving fastq data.
We will begin with downloading the data off NCBI Gene Expression Omnibus. This will download all fastq files of cells from E14.5 timepoint
```{bash}
cd /media/emlab/RAID5/Fursham/Oscar_interneuron
mkdir fastq
cd fastq
while IFS=$'\t' read -r SRA ID TYPE TYPEID STATE STATEID CLUSTERID;do 
	fastq-dump --gzip $SRA 
done < ~/Documents/git_repo/AS-scRNAseq/cluster_information.txt
```


## 3. Running whippet quantification
We will merge fastq files from cells based on its cell type and align that merged file to mm10 genome. Whippet indices for mm10 genome have been prebuilt and found at /media/emlab/RAID5/Genomes....

```{bash}
cd /media/emlab/RAID5/Fursham/Oscar_interneuron
mkdir -p whippet/quantify
mkdir whippet_SAMs

# merge by cell state
declare -A whippetArray
declare -A whippetNames
while IFS=$'\t' read -r SRA ID TYPE TYPEID STATE STATEID CLUSTERID;do 
  key="$STATEID"
  whippetArray[$key]="fastq/$SRA.fastq.gz ${whippetArray[$key]}"
  whippetNames[$key]=$STATE
done < ~/Documents/git_repo/AS-scRNAseq/cluster_information.txt

#run whippet
for (( n=1; n<=2; n++ ));do
	whippet-quant.jl <( cat ${whippetArray[$n]} ) -o whippet/quantify/${whippetNames["$n"]} -x /media/emlab/RAID5/Genomes/GRCm38_mm10/Ensembl/whippet_index/graph.jls --force-gz --biascorrect --sam > whippet_SAMs/${whippetNames[$n]}.sam 
done

# merge by cell type and proliferative state
declare -A whippetArray2
declare -A whippetNames2
while IFS=$'\t' read -r SRA ID TYPE TYPEID STATE STATEID CLUSTERID;do 
  key="$STATEID.$TYPEID"
  whippetArray2[$key]="fastq/$SRA.fastq.gz ${whippetArray2[$key]}"
  whippetNames2[$key]="$STATE-$TYPE"
done < ~/Documents/git_repo/AS-scRNAseq/cluster_information.txt

#run whippet
for id in ${!whippetArray2[@]};do
	whippet-quant.jl <( cat ${whippetArray2[$id]} ) -o whippet/quantify/${whippetNames2["$id"]} -x /media/emlab/RAID5/Genomes/GRCm38_mm10/Ensembl/whippet_index/graph.jls --force-gz --biascorrect --sam > whippet_SAMs/${whippetNames2[$id]}.sam 
done


# merge by cell type only
declare -A whippetArray3
declare -A whippetNames3
declare -A whippetArray4
declare -A whippetNames4
while IFS=$'\t' read -r SRA ID TYPE TYPEID STATE STATEID CLUSTERID;do 
  whippetArray3[$TYPE]="fastq/$SRA.fastq.gz ${whippetArray3[$TYPE]}"
  whippetNames3[$TYPE]="$TYPE"
  whippetArray4[$CLUSTERID]="fastq/$SRA.fastq.gz ${whippetArray4[$CLUSTERID]}"
  whippetNames4[$CLUSTERID]="Cluster_$CLUSTERID"
done < ~/Documents/git_repo/AS-scRNAseq/cluster_information.txt

#run whippet
for id in dMGE vMGE CGE;do
	whippet-quant.jl <( cat ${whippetArray3[$id]} ) -o whippet/quantify/${whippetNames3["$id"]} -x /media/emlab/RAID5/Genomes/GRCm38_mm10/Ensembl/whippet_index/graph.jls --force-gz --biascorrect --sam > whippet_SAMs/${whippetNames3[$id]}.sam 
done

for id in 0 2;do
	whippet-quant.jl <( cat ${whippetArray4[$id]} ) -o whippet/quantify/${whippetNames4["$id"]} -x /media/emlab/RAID5/Genomes/GRCm38_mm10/Ensembl/whippet_index/graph.jls --force-gz --biascorrect --sam > whippet_SAMs/${whippetNames4[$id]}.sam 
done

```

## 4. Running whippet differential splicing
We will now compare differential splicing between the comparisons mentioned on line 6  
We will start by preparing a data-frame of containing a set of samples to compare
```{r}
setwd("/media/emlab/RAID5/Fursham/Oscar_interneuron")
library(tidyverse)
tibble::tibble(sampleA = c("Proliferative", "Proliferative-CGE", "Proliferative-dMGE", "Proliferative-vMGE", "Proliferative-vMGE", "Proliferative-vMGE", "Proliferative-dMGE", "Non-proliferative-vMGE", "Non-proliferative-vMGE", "Non-proliferative-dMGE"),
               sampleB = c("Non-proliferative", "Non-proliferative-CGE", "Non-proliferative-dMGE", "Non-proliferative-vMGE", "Proliferative-dMGE", "Proliferative-CGE", "Proliferative-CGE", "Non-proliferative-dMGE", "Non-proliferative-CGE", "Non-proliferative-CGE"),
               name = c("Proliferative_vs_Non-proliferative", "Proliferative_vs_Non-proliferative_CGE", "Proliferative_vs_Non-proliferative_dMGE", "Proliferative_vs_Non-proliferative_vMGE", "vMGE_vs_dMGE_Proliferative", "vMGE_vs_CGE_Proliferative", "dMGE_vs_CGE_Proliferative", "vMGE_vs_dMGE_Non-proliferative", "vMGE_vs_CGE_Non-proliferative", "dMGE_vs_CGE_Non-proliferative")) %>% write.table("whippet_comparisons.txt", sep = "\t", quote = F, row.names = F, col.names = F)
  
```


Using the list of comparisons, run whippet delta splicing analysis

```{bash}
cd /media/emlab/RAID5/Fursham/Oscar_interneuron/whippet/quantify
mkdir ../delta
while IFS=$'\t' read -r A B NAME;do 
  whippet-delta.jl -a $A.psi.gz, -b $A.psi.gz, -o ../delta/$NAME
done < ../../whippet_comparisons.txt

# compare cell types
whippet-delta.jl -a vMGE.psi.gz, -b dMGE.psi.gz, -o ../delta/vMGE_dMGE
whippet-delta.jl -a vMGE.psi.gz, -b CGE.psi.gz, -o ../delta/vMGE_CGE
whippet-delta.jl -a dMGE.psi.gz, -b CGE.psi.gz, -o ../delta/dMGE_CGE
whippet-delta.jl -a Cluster_0.psi.gz, -b Cluster_2.psi.gz, -o ../delta/Cluster0_Cluster2

```

## 4. Convert aligned sequence file (SAMs) to binary format (BAMs) and index them
```{bash}
cd /media/emlab/RAID5/Fursham/Oscar_interneuron
mkdir whippet_sorted_BAMs
for filename in whippet_SAMs/*.sam; do
	file=${filename##*/}
	file=${file%.sam}
	samtools view -@ 40 -Su $filename | samtools sort -@ 40 -o whippet_sorted_BAMs/$file.bam
done



# 5) generate indexed bamfiles
cd /media/emlab/RAID5/Fursham/Yeo_sALS/whippet_sorted_BAMs
for filename in *.bam; do
    samtools index -@ 40 $filename $filename.bai
done
```


